File: src/launch.rs
use std::fs;
use std::os::unix::process::CommandExt;
use std::process::{Command, Stdio};

pub fn game_pid_file() -> String {
    std::env::var("ARKANA_GAME_PIDFILE").unwrap_or_else(|_| "/run/arkana-game.pid".to_string())
}

pub fn launch_game(exec: &str) -> bool {
    let launcher_pid = std::process::id();
    let script = format!("while kill -0 {launcher_pid} 2>/dev/null; do sleep 0.2; done\n{exec}");

    let devnull_in = fs::File::open("/dev/null").ok();
    let devnull_out = fs::OpenOptions::new().write(true).open("/dev/null").ok();
    let devnull_err = fs::OpenOptions::new().write(true).open("/dev/null").ok();

    let mut cmd = Command::new("sh");
    cmd.args(["-c", &script]);

    if let Some(f) = devnull_in {
        cmd.stdin(Stdio::from(f));
    }
    if let Some(f) = devnull_out {
        cmd.stdout(Stdio::from(f));
    }
    if let Some(f) = devnull_err {
        cmd.stderr(Stdio::from(f));
    }

    unsafe {
        cmd.pre_exec(|| {
            libc::setsid();
            Ok(())
        });
    }

    match cmd.spawn() {
        Ok(child) => {
            let pid = child.id();
            let pidfile = game_pid_file();
            let _ = fs::write(pidfile, pid.to_string());
            true
        }
        Err(err) => {
            eprintln!("launch_game: spawn failed: {err}");
            false
        }
    }
}

File: src/launch.cpp
#include "launch.h"

#include <cerrno>
#include <cstdio>
#include <filesystem>
#include <format>
#include <fstream>
#include <iostream>
#include <print>
#include <string>

#include <fcntl.h>
#include <unistd.h>

#include "unique_fd.h"

static std::string game_pid_file() {
    const char* env = std::getenv("ARKANA_GAME_PIDFILE");
    if (env && env[0]) {
        return std::string(env);
    }
    return "/run/arkana-game.pid";
}

bool launch_game(const std::string& exec) {
    pid_t launcher_pid = getpid();

    auto script = std::format("while kill -0 {} 2>/dev/null; do sleep 0.2; done\n{}",
                              launcher_pid, exec);

    pid_t child = fork();
    if (child < 0) {
        std::println(std::cerr, "launch_game: fork failed: {}", std::format("errno {}", errno));
        return false;
    }

    if (child == 0) {
        setsid();
        int devnull_fd = open("/dev/null", O_RDWR);
        UniqueFd devnull(devnull_fd);
        if (devnull.valid()) {
            dup2(devnull.get(), STDIN_FILENO);
            dup2(devnull.get(), STDOUT_FILENO);
            dup2(devnull.get(), STDERR_FILENO);
        }
        execlp("sh", "sh", "-c", script.c_str(), nullptr);
        _exit(127);
    }

    std::string pidfile = game_pid_file();
    std::ofstream pf;
    pf.open(pidfile);
    if (pf.is_open()) {
        pf << child;
    }

    return true;
}
